openstack: Use common SafeFormatAndMount() for mounting.

jsafrane opened the pull request:
There is no reason for OpenStack + Cinder to use it's own variant of format
& mount.

Fixes #17565.

swagiaal commented on pkg/volume/cinder/cinder.go:
```diff
@@ -89,7 +89,7 @@ func (plugin *cinderPlugin) newBuilderInternal(spec *volume.Spec, podUID types.U
 		fsType:             fsType,
 		readOnly:           readOnly,
-		blockDeviceMounter: &cinderSafeFormatAndMount{mounter, exec.New()}}, nil
+		blockDeviceMounter: &mount.SafeFormatAndMount{plugin.host.GetMounter(), exec.New()}}, nil
```
Hmm.. s/plugin.host.GetMounter()/mounter/
Not sure if the cinder tests use this but other volume plugin tests use this to pass in mock arguments

jsafrane replied:
Fixed. (Still, AWS has the same `plugin.host.GetMounter()` call)

swagiaal replied:
> Fixed. 

Great.. thanks

>(Still, AWS has the same plugin.host.GetMounter() call)

Ah! True...  That is because AWS and Cinder tests do not have this check:
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/gce_pd_test.go#L169 :(

swagiaal commented:
@k8s-bot unit test this

swagiaal commented:
@jsafrane could you please rebase on master

jsafrane commented:
Rebased

swagiaal commented on pkg/api/validation/validation.go:
```diff
@@ -606,8 +606,8 @@ func validateCinderVolumeSource(cd *api.CinderVolumeSource) validation.ErrorList
-	if cd.FSType == "" || (cd.FSType != "ext3" && cd.FSType != "ext4") {
-		allErrs = append(allErrs, validation.NewFieldRequired("fsType required and should be of type ext3 or ext4"))
+	if cd.FSType == "" {
+		allErrs = append(allErrs, validation.NewFieldRequired("fsType"))
```
I just noticed this. fsType is not required. mount.SafeFormatAndMount will default to ext4. In fact the parameter has "omitempty" in `types.go` even though the comment says required.

jsafrane replied:
I would do that in separate PR, it's 'omitempty' in almost all volume plugins, while the validator for all of them requires it's presence.

swagiaal replied:
> I would do that in separate PR, it's 'omitempty' in almost all volume plugins, while the validator for all of them requires it's presence.

Ok.. sounds good

swagiaal commented on pkg/api/types.go:
```diff
@@ -575,7 +575,6 @@ type CinderVolumeSource struct {
 	VolumeID string `json:"volumeID"`
 	// Required: Filesystem type to mount.
```
Could you remove the comment above as fsType is not required.

jsafrane replied:
Done.

swagiaal commented:
Just a couple of nits but otherwise LGTM

jsafrane commented:
Another rebase.

jsafrane commented:
@k8s-bot test this please

swagiaal commented:
@k8s-bot unit test this

ncdc commented:
@k8s-bot unit test this

ncdc commented:
@jsafrane the build log indicates a compilation failure:

```
_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder_util_test.go:60: undefined: cinderSafeFormatAndMount
```

jsafrane commented:
@k8s-bot unit test this

jsafrane commented:
@k8s-bot test this

swagiaal commented:
hmm.. I did not realize this still did not get merged :(
@jsafrane could you rebase please

jsafrane committed:
openstack: Use common SafeFormatAndMount() for mounting.

There is no reason for OpenStack + Cinder to use it's own variant of format
& mount.

jsafrane commented:
Rebased.

swagiaal commented:
@jsafrane thanks
@markturansky can you take a look at this ? it LGTM.

markturansky commented:
This LGTM.  @jsafrane or @swagiaal, let me know when the local testing is done (I know Jan did it quite some time ago) and I'll add the label.

jsafrane commented:
I tested it with today's master and there was actually a bug, `cinder_util.go` should use `FormatAndMount` instead of plain `Format`:

```diff
--- a/pkg/volume/cinder/cinder_util.go
+++ b/pkg/volume/cinder/cinder_util.go
@@ -80,7 +79,7 @@ func (util *CinderDiskUtil) AttachDisk(b *cinderVolumeBuilder, globalPDPath stri
                }
        }
        if notmnt {
-               err = b.blockDeviceMounter.Mount(devicePath, globalPDPath, b.fsType, options)
+               err = b.blockDeviceMounter.FormatAndMount(devicePath, globalPDPath, b.fsType, options)
                if err != nil {
                        os.Remove(globalPDPath)
                        return err

```

I pushed fixed patch and tested that Cinder volumes with `FsType="xfs"` are now automatically formatted with `xfs`, which was not possible before this patch.

markturansky commented:
@jsafrane thanks, Jan.  Would that bug have caught by the unit test that was removed?  The test was originally for its own mounter, but could it be modified to use the mounter you're replacing it with?

jsafrane commented:
The test tested the custom mounter I removed, not the plugin itself.

markturansky commented:
Understood.  I think a unit test for the current mounter would be helpful, too.

markturansky commented:
Thanks.  I'll have the confidence for LGTM with a unit test for the mounter.

jsafrane commented:
OpenStack.AttachDisk is directly called from the cinder_util.go, so the test could run only with real OpenStack. No volume plugin is decoupled this way, unit testing of *_util.go files is not easily possible... We should make it better when rewriting the plugins for Attach controller.

swagiaal commented:
Also, SafeFormatAndMount is fairly well unit tested independently of the plugins that use it.

markturansky commented:
Ok.  SafeFormatAndMount certainly is unit tested and tested daily by everyone that mounts volumes.  Moving to it from a customer Cinder mounter is a good improvement.

swagiaal commented:
@markturansky thanks for reviewing. @jsafrane good catch on that bug I can't believe I missed it having written the PR to rename Mount to FormatAndMount :/
